src/libsync/abstractcorejob.h

```diff
@@ -0,0 +1,69 @@
+#pragma once
+
+#include <QNetworkAccessManager>
+#include <QNetworkReply>
+
+#include "owncloudlib.h"
+
+namespace OCC {
+
+class OWNCLOUDSYNC_EXPORT Job : public QObject
+{
+    Q_OBJECT
+
+    friend class AbstractCoreJobFactory;
+
+public:
+    explicit Job(QObject *parent = nullptr);
+
+    [[nodiscard]] const QVariant &result() const;
+
+    [[nodiscard]] const QString &errorMessage() const;
+    [[nodiscard]] QNetworkReply::NetworkError networkError() const;
+
+    [[nodiscard]] bool success() const;
+
+protected:
+    void finishWithResult(const QVariant &result);
+
+    void finishWithError(const QString &errorMessage, QNetworkReply::NetworkError networkError);
+
+Q_SIGNALS:
+    void finished();
+
+private:
+    bool _success = false;
+    QVariant _result;
+
+    QString _errorMessage;
+    QNetworkReply::NetworkError _networkError = QNetworkReply::NoError;
+};
+
+class OWNCLOUDSYNC_EXPORT AbstractCoreJobFactory : public QObject
+{
+    Q_OBJECT
+
+public:
+    explicit AbstractCoreJobFactory(QNetworkAccessManager *nam, QObject *parent = nullptr);
+    ~AbstractCoreJobFactory() override;
+
+    virtual Job *startJob(const QUrl &url) = 0;
+
+protected:
+    [[nodiscard]] QNetworkAccessManager *nam() const;
+
+    static void finishJobWithSuccess(Job *job, const QVariant &result)
```

Comment:
Maybe also set*?

Reply:
We should reconsider removing the `friend` class stuff and using public setters IMO.